fem/fespace.cpp

```diff
@@ -1632,167 +1879,404 @@ void FiniteElementSpace::Construct()
+         nfdofs = MakeDofTable(2, face_orders, var_face_dofs,
+                               IsVariableOrder() ? &var_face_orders : NULL);
+         uni_fdof = -1;
+      }
+      else
+      {
+         // the simple case: all faces are of the same geometry and order
+         uni_fdof = fec->GetNumDof(mesh->GetFaceGeometry(0), order);
+         nfdofs = mesh->GetNFaces() * uni_fdof;
+      }
+   }
+
+   // assign internal ("bubble") DOFs
+   if (mesh->GetNE() && dim > 0)
+   {
+      if (IsVariableOrder() || mixed_elements)
+      {
+         bdofs = new int[mesh->GetNE()+1];
+         bdofs[0] = 0;
+         for (int i = 0; i < mesh->GetNE(); i++)
+         {
+            int p = GetElementOrderImpl(i);
+            nbdofs += fec->GetNumDof(mesh->GetElementGeometry(i), p);
+            bdofs[i+1] = nbdofs;
+         }
+      }
+      else
+      {
+         // the simple case: all elements are the same
+         bdofs = NULL;
+         Geometry::Type geom = mesh->GetElementGeometry(0);
+         nbdofs = mesh->GetNE() * fec->GetNumDof(geom, order);
       }
    }
 
    ndofs = nvdofs + nedofs + nfdofs + nbdofs;
 
+   // DOFs are now assigned according to current element orders
+   orders_changed = false;
+
    // Do not build elem_dof Table here: in parallel it has to be constructed
    // later.
 }
 
-void FiniteElementSpace::GetElementDofs(int i, Array<int> &dofs) const
+int FiniteElementSpace::MinOrder(VarOrderBits bits)
 {
-   if (elem_dof)
+   MFEM_ASSERT(bits != 0, "invalid bit mask");
+   for (int order = 0; bits != 0; order++, bits >>= 1)
    {
-      elem_dof -> GetRow (i, dofs);
+      if (bits & 1) { return order; }
    }
-   else
+   return 0;
+}
+
+void FiniteElementSpace
+::CalcEdgeFaceVarOrders(Array<VarOrderBits> &edge_orders,
+                        Array<VarOrderBits> &face_orders) const
+{
+   MFEM_ASSERT(IsVariableOrder(), "");
+   MFEM_ASSERT(Nonconforming(), "");
+   MFEM_ASSERT(elem_order.Size() == mesh->GetNE(), "");
+
+   edge_orders.SetSize(mesh->GetNEdges());  edge_orders = 0;
+   face_orders.SetSize(mesh->GetNFaces());  face_orders = 0;
+
+   // initial edge/face orders, as required by incident elements
+   Array<int> E, F, ori;
+   for (int i = 0; i < mesh->GetNE(); i++)
    {
-      Array<int> V, E, Eo, F, Fo;
-      int k, j, nv, ne, nf, nb, nfd, nd, dim;
-      const int *ind;
+      int order = elem_order[i];
+      MFEM_ASSERT(order <= MaxVarOrder, "");
+      VarOrderBits mask = (VarOrderBits(1) << order);
 
-      dim = mesh->Dimension();
-      nv = fec->DofForGeometry(Geometry::POINT);
-      ne = (dim > 1) ? ( fec->DofForGeometry(Geometry::SEGMENT) ) : ( 0 );
-      nb = (dim > 0) ? fec->DofForGeometry(mesh->GetElementBaseGeometry(i)) : 0;
-      if (nv > 0)
+      mesh->GetElementEdges(i, E, ori);
+      for (int j = 0; j < E.Size(); j++)
       {
-         mesh->GetElementVertices(i, V);
+         edge_orders[E[j]] |= mask;
```

Comment:
Some comments describing the purpose of this bitmask operation would be appreciated.

Reply:
Added.